### PR TITLE
Allow users to hide global announcements they have read

### PR DESCRIPTION
**Pull request by kxmbrian**
For #597 

# Previously
screen shot 2016-07-04 at 12 40 35 pm
- Can close the panel, but it reappears when you refresh or go to another page. In effect, it acts like a sticky announcement.
- Only shows latest announcement. "More Announcements..." link is shown if there are any other active global announcements.

# Now
screen shot 2016-07-04 at 12 25 19 pm
- Closed announcement panels will not reappear
- Since they can be closed, better to show all unread announcements rather than have the user see the next announcement only upon refreshing the page or by going to the global announcements index page.
- Once merged, need to raise issue to provide link to the global announcements index page so that user can view active announcements that he has closed.
- Even when the user has visited the global announcements index page, she will still have to mark each one as read by closing the panels. Don't want to just mark all as read when the user requests the index page since the server cannot guarantee that nothing will go wrong after the response has been sent, i.e. cannot guarantee user will read the announcements.

### CODE REVIEW

File: app/controllers/announcements_controller.rb [hunk 1]
```diff
@@ -1,8 +1,17 @@
...
 
-  def index
+  def index #:nodoc:
     @announcements = global_announcements.includes(:creator)
   end
+
+  # Marks the current GenericAnnouncement as read by the current user and responds without a body.
+  def mark_as_read
+    @announcement.mark_as_read! for: current_user
+    head :ok
```
Comment (kxmbrian) [inline app/controllers/announcements_controller.rb]: @allenwq @weiqingtoh I don't think authorization is necessary here. Also, I'm not sure if I need to handle the case where `mark_as_read!` fails. Any thoughts?

Comment (allenwq) [inline app/controllers/announcements_controller.rb]: Yes agree. And if it's not supposed to fail, just raise an exception when fail.

File: app/controllers/announcements_controller.rb [hunk 2]
```diff
@@ -1,8 +1,17 @@
 # frozen_string_literal: true
 class AnnouncementsController < ApplicationController
+  load_resource :announcement, class: GenericAnnouncement.name, only: :mark_as_read,
+                               id_param: :announcement_id
+
   add_breadcrumb :index, :announcements_path
 
-  def index
+  def index #:nodoc:
```
Comment (allenwq) [inline app/controllers/announcements_controller.rb]: Actually don't need to `:nodoc:` anymore.

Comment (kxmbrian) [inline app/controllers/announcements_controller.rb]: Is it okay if I leave it there this time?

File: app/controllers/announcements_controller.rb (hunk 1 above)
Comment (weiqingtoh) [inline app/controllers/announcements_controller.rb]: @allenwq so in this case this is enough right?

Comment (kxmbrian) [inline app/controllers/announcements_controller.rb]: `mark_as_read!` [returns](https://github.com/ledermann/unread/blob/master/lib/unread/readable.rb) the result of a call to `transaction`, which should raise an exception if something goes bad.

File: app/controllers/announcements_controller.rb (hunk 2 above)
Comment (allenwq) [inline app/controllers/announcements_controller.rb]: Sure.

File: app/controllers/announcements_controller.rb (hunk 1 above)
Comment (allenwq) [inline app/controllers/announcements_controller.rb]: @weiqingtoh @kxmbrian Yes, so it's enough.

Comment (lowjoel) [inline app/controllers/announcements_controller.rb]: `head` would give no body. Is the mark as read link/button done using script?

File: config/routes.rb [hunk 3]
```diff
@@ -73,7 +73,9 @@
     masquerades: 'user/masquerades'
   }
 
-  resources :announcements, only: [:index]
+  resources :announcements, only: [:index] do
+    patch 'mark_as_read'
```
Comment (lowjoel) [inline config/routes.rb]: `POST` makes sense here.

File: app/controllers/announcements_controller.rb (hunk 2 above)
Comment (kxmbrian) [inline app/controllers/announcements_controller.rb]: I've removed `:nodoc:` in the latest update

**Comment by weiqingtoh** [comment]
👍 LGTM

**Pull request merged**